numpyro/infer/elbo.py

```diff
@@ -577,34 +543,47 @@ def single_particle_elbo(rng_key):
             check_model_guide_match(model_trace, guide_trace)
             _validate_model(model_trace, plate_warning="strict")
 
-            # XXX: different from Pyro, we don't support baseline_loss here
-            non_reparam_nodes = {
-                name
-                for name, site in guide_trace.items()
-                if site["type"] == "sample"
-                and (not site["is_observed"])
-                and (not site["fn"].has_rsample)
-            }
-            if non_reparam_nodes:
-                downstream_costs, _ = _compute_downstream_costs(
-                    model_trace, guide_trace, non_reparam_nodes
+            # Find dependencies on non-reparameterizable sample sites for
+            # each cost term in the model and the guide.
+            model_deps, guide_deps = get_provenance(
+                eval_provenance(
+                    partial(
+                        track_nonreparam(get_importance_log_probs),
+                        seeded_model,
+                        seeded_guide,
+                        args,
+                        kwargs,
+                    )
```

Comment:
For test, we can do the traditional provenance tracking to get direct dependencies, then do postprocessing and compare with this implementation.